### DISCUSSION TITLE
Extend from IDbConnection instead

### DISCUSSION THREAD

**Discussion by goncalo-oliveira** [vb-consulting/Norm.net#13] [Ideas]
Firstly, congrats on the library. With the current *state* of Dapper, this is a pretty interesting alternative for those that don't like EF performance (or philosophy).

I'd like to ask you why did you choose to implement the extensions on top of `System.Data.Common.DbConnection` component instead of the `System.Data.IDbConnection` interface.

**Comment by vbilopav** [comment]
Thank you.

I honestly can't remember from the top of my head, it was some time ago. I know that I started with IDbConnection but I reverted to DbConnection for some reason. 

I'll need to check it a bit deeper to give you an answer. Maybe it can be reverted to IDbConnection in future, but at this moment I don't have any plans of doing new versions since this one serves me well.

**Reply by goncalo-oliveira** [reply]
Possibly because of the async API that isn't present in the interface. I have a library for having a DI compatible factory that works around `IDbConnection`. I think the reason at the time was to support ADO drivers that implemented directly the interface instead of the abstract class.
I remember that some async methods are wrapped around code that checks whether we have a `DbConnection` instance that natively supports the async call or creates some sort of adapter that can either use the native call if it exists (reflection) or wrap the sync call with async code.
I'm unsure if today there are still drivers where we don't have a `DbConnection` instance; at least Microsoft's SqlClient and Npgsql do, but there are others out there...
I think Dapper had the same discussion a few years ago and eventually went with the interface.

Anyway... for my current scenario, I can do with the library as it is, since the use case uses Microsoft's SqlClient. But maybe worth discussing for a future version...

Thanks for replying!

**Reply by vbilopav** [reply]
I'll look it up certainly. Maybe I could support both.

**Comment by vbilopav** [comment]
Edit: I'll be looking into this issue shortly: https://github.com/vb-consulting/Norm.net/issues/14 and I'll try to include this support in this fix.

**Comment by vbilopav** [comment]
Hi @goncalo-oliveira 

I'm just preparing to release a new version. I've tested the library with the `Microsoft.Data.SqlClient` and it works just the same as with `System.Data.SqlClient`.

You can see some tests here:
https://github.com/vb-consulting/Norm.net/blob/5.3.2/Tests/SqlServerUnitTests/MicrosoftClientUnitTests.cs

Can you try to describe the issue so I can include it in this release?

Thanks
